core/src/main/java/org/apache/calcite/rel/rules/PruneEmptyRules.java

```diff
@@ -540,4 +540,32 @@ public interface JoinRightEmptyRuleConfig extends PruneEmptyRule.Config {
       };
     }
   }
+
+
+  public static final RelOptRule EMPTY_TABLE =
+      ImmutableEmptyTableOptimizationConfig.of()
+          .withOperandSupplier(b0 -> {
+            return b0.operand(TableScan.class).noInputs();
+          })
+          .withDescription("ConvertEmptyTableToValues")
+          .toRule();
+
+  /** Configuration for rule that transforms an empty table into an empty values node.
+   * MaxRowCount is used as the stat to transform, hence Table implementer needs
+   * to supply this metadata if this optimization needs to be applied.*/
+  @Value.Immutable
+  public interface EmptyTableOptimizationConfig extends PruneEmptyRule.Config {
+
+    @Override default PruneEmptyRule toRule() {
+      return new PruneEmptyRule(this) {
+        @Override public void onMatch(RelOptRuleCall call) {
+          TableScan tableScan = call.rel(0);
```

Comment:
It is better to be more general here (use `RelNode` here instead of `TableScan`); the rule will still work correctly and if people want to reduce or enlarge the scope they can do it by changing the matching operands of the configuration.